library/src/main/java/com/datatorrent/lib/state/managed/BucketsFileSystem.java

```diff
@@ -0,0 +1,580 @@
+        }
+      }
+      return immutableTimeBucketMetas;
+    }
+  }
+
+  /**
+   * Loads the bucket meta-file.
+   *
+   * @param bucketId bucket id
+   * @param dis      data input stream
+   * @throws IOException
+   */
+  private void loadBucketMetaFile(long bucketId, DataInputStream dis) throws IOException
+  {
+    int numberOfEntries = dis.readInt();
+
+    for (int i = 0; i < numberOfEntries; i++) {
+      long timeBucketId = dis.readLong();
+      long dataSize = dis.readLong();
+      long lastTransferredWindow = dis.readLong();
+
+      MutableTimeBucketMeta tbm = new MutableTimeBucketMeta(bucketId, timeBucketId);
+
+      int sizeOfFirstKey = dis.readInt();
+      byte[] firstKeyBytes = new byte[sizeOfFirstKey];
+      dis.readFully(firstKeyBytes, 0, firstKeyBytes.length);
+      tbm.updateTimeBucketMeta(lastTransferredWindow, dataSize, new Slice(firstKeyBytes));
+
+      timeBucketsMeta.put(bucketId, timeBucketId, tbm);
+    }
+  }
+
+  /**
+   * Saves the updated bucket meta on disk.
+   *
+   * @param bucketId bucket id
+   * @throws IOException
+   */
+  void updateBucketMetaFile(long bucketId) throws IOException
+  {
+    Map<Long, MutableTimeBucketMeta> timeBuckets;
+    synchronized (timeBucketsMeta) {
+      timeBuckets = timeBucketsMeta.row(bucketId);
+    }
+    Preconditions.checkNotNull(timeBuckets, "timeBuckets");
+    String tmpFileName = getTmpFileName();
+
+    try (DataOutputStream dos = getOutputStream(bucketId, tmpFileName)) {
+      dos.writeInt(timeBuckets.size());
+      for (Map.Entry<Long, MutableTimeBucketMeta> entry : timeBuckets.entrySet()) {
+        MutableTimeBucketMeta tbm = entry.getValue();
+        dos.writeLong(tbm.getTimeBucketId());
+        dos.writeLong(tbm.getSizeInBytes());
+        dos.writeLong(tbm.getLastTransferredWindowId());
+        dos.writeInt(tbm.getFirstKey().length);
+        dos.write(tbm.getFirstKey().toByteArray());
+      }
+
+    }
+    rename(bucketId, tmpFileName, META_FILE_NAME);
+  }
+
+  protected void deleteTimeBucketsLessThanEqualTo(long latestExpiredTimeBucket) throws IOException
+  {
+    LOG.debug("delete files before {}", latestExpiredTimeBucket);
+
+    for (long bucketName : bucketNamesOnFS) {
+      RemoteIterator<LocatedFileStatus> timeBucketsIterator = listFiles(bucketName);
+      boolean emptyBucket = true;
+      while (timeBucketsIterator.hasNext()) {
+        LocatedFileStatus timeBucketStatus = timeBucketsIterator.next();
+
+        String timeBucketStr = timeBucketStatus.getPath().getName();
+        if (timeBucketStr.equals(BucketsFileSystem.META_FILE_NAME) || timeBucketStr.endsWith(".tmp")) {
+          //ignoring meta and tmp files
+          continue;
+        }
+        long timeBucket = Long.parseLong(timeBucketStr);
+
+        if (timeBucket <= latestExpiredTimeBucket) {
+          LOG.debug("deleting bucket {} time-bucket {}", timeBucket);
+          delete(bucketName, timeBucketStatus.getPath().getName());
+
```

Comment:
The time bucket data file is deleted then the meta file is updated. If the operator fails after deleting the time bucket but before the meta file is updated, the operator could load the meta data for a deleted time bucket on recovery. Could that cause a problem? Could the update of the Meta file be done before the delete of the time bucket?